[FIX] website: enable to use the facebook snippet inside a TOC

loco-odoo opened the pull request:
Steps to reproduce the bug:
- Drop a "Table of Content" snippet inside a page.
- Drop a "Facebook" snippet inside the "Table of Content".
-> The page is blocked.

A mechanism that interests us here is the mutation observer. Up to this
commit its goal was to regenerate the navbar of the TOC when its DOM
content changed. In practice, since [1], a "widgets_start_request"
message was triggered up each time there was a change of the TOC
content. In the case of the Facebook snippet, this behavior led to an
infinite loop. Indeed, when the Facebook snippet was dropped inside a
TOC, its content changed so a "widgets_start_request" was triggered up.
This message led to the execution of the start function of the TOC but
also of the start function of the Facebook snippet. Because there is a
change of the DOM in this last function, the mutation observer
triggered up a "widgets_start_request" message again that caused the
program to loop forever.

The goal of this commit is to regenerate the content of the navbar only
when its content changes. To do so, the content of the navbar before
the change of the DOM is compared with its content after the change of
the DOM. If those two contents are different, it means that the navbar
should be updated. Thanks to this commit, it is now possible to add
dynamic snippets such as "Facebook" or "Countdown" inside the TOC.

With this in hands, another bug appeared:
- Drop a "Table of Content" snippet inside a page.
- Drop a "Text-Image" snippet on the page but not inside the TOC.
- Drop a "Countdown" snippet inside the "Text-Image".
- Drag the entire "Text-Image" snippet with "Countdown" and drop it
inside the TOC.
-> The page is blocked.

This bug comes from the fact that the drop of a snippet that stands on
the page inside an other snippet leads to the destroy and the restart
of the option of the outer snippet. However, in our case, the mutation
observer was still listening to DOM changes even after the destroy of
the snippet option. This led to parasite function call. This bug is
fixed by disconnecting the mutation observer when the option is
destroyed. Note that this is done in this PR ([2]) as the problem
appeared in older version as well.

[1]: https://github.com/odoo/odoo/commit/a05f782871c61b2d58ca2fd4277cb7aed65a301d
[2]: https://github.com/odoo/odoo/pull/112960

task-3122249
opw-3164969
opw-3173006

qsm-odoo commented on addons/website/static/src/snippets/s_table_of_content/options.js:
```diff
@@ -69,23 +78,42 @@ options.registry.TableOfContent = options.Class.extend({
+        }
+    },
+
+    /**
+    * The navbar needs to be updated if its content before the change of the
+    * DOM is different fom the one after the change of the DOM.
+    */
+    _navbarNeedUpdate(oldHeadingsEls, newHeadingsEls) {
```
- Wrong indentation of the JS DOC. Try to install/fix something on your editor so it is correctly indented automatically :+1: 
- I think it deserves a bit more explanation why this is necessary (either here or the caller). IMO you should not make a function out of this you only call it once and it is very small. A computation of a variable `areHeadingsEqual` and a commented `if (!areHeadingsEqual)` would be way clearer IMO.
- If we would keep this function: `_navbarNeedUpdate` -> `_navbarNeedsUpdate` (but let's not keep it)

qsm-odoo reviewed (changes requested):
Good catch and good explanation :+1: Some requested changes.

loco-odoo commented:
@qsm-odoo, I just modified the PR as you asked it and I created an other one (https://github.com/odoo/odoo/pull/112960/commits/d35cff1ee66a9dba2d030f2d516b4c9cbc17e6c5) for the memory leak (I also modified the commit message of this PR to mention it) :+1:

qsm-odoo reviewed (approved):
I added [this diff](https://github.com/odoo/odoo/compare/8fcbdc266e9ddd68c35550644f5a78000e32c11c..0a156084801708936b9ecbaf1ccf03c40202d177) (it's actually the important info I expected as a comment: with only your comment I could say "Mmh, ok, but we could just regenerate it everytime, that would be less LOC").

Good addition to the commit message, I actually did not understand that the second steps were another usecase that is broken because of the missing disconnect in the destroy :+1: 

@robodoo r+
